src/crimson/osd/pg.h

```diff
@@ -540,6 +541,20 @@ class PG : public boost::intrusive_ref_counter<
     eversion_t &version);
 
 private:
+
+  struct SnapTrimMutex {
+    struct WaitPG : OrderedConcurrentPhaseT<WaitPG> {
+      static constexpr auto type_name = "SnapTrimEvent::wait_pg";
+    } wait_pg;
+    seastar::shared_mutex mutex;
+
+    interruptible_future<> lock(SnapTrimEvent &st_event) noexcept;
+
+    void unlock() noexcept {
+      mutex.unlock();
+    }
+  } snaptrim_mutex;
+
```

Comment:
Can this be moved to  snaptrim_event header? The other phases (`WaitSubop` and `WaitTrimTimer`) are already declared there.

---

Instead of using an additional mutex, can we re-use the mutex used from within `OrderedExclusivePhaseT`?
`OrderedExclusivePhaseT` and `OrderedConcurrentPhaseT` use their own mutexes to ensure order or/and exclusiveness.
IIUC, `OrderedExclusivePhaseT` can be used here instead to ensure no concurrent trimming is happening.

OrderedExclusivePhaseT:
```
/**
 * Ensures that at most one op may consider itself in the phase at a time.
 * Ops will see enter() unblock in the order in which they tried to enter
 * the phase.  entering (though not necessarily waiting for the future to
 * resolve) a new phase prior to exiting the previous one will ensure that
 * the op ordering is preserved.
 */
```

Reply:
> Can this be moved to snaptrim_event header? The other phases (`WaitSubop` and `WaitTrimTimer`) are already declared there.

I think snaptrim_mutex has to be a member of PG, because the synchonization of SnapTrimEvents happens within the scope of PG. Am I missing any thing?

> 
> Instead of using an additional mutex, can we re-use the mutex used from within `OrderedExclusivePhaseT`? `OrderedExclusivePhaseT` and `OrderedConcurrentPhaseT` use their own mutexes to ensure order or/and exclusiveness. IIUC, `OrderedExclusivePhaseT` can be used here instead to ensure no concurrent trimming is happening.
> 
> OrderedExclusivePhaseT:
> 
> ```
> /**
>  * Ensures that at most one op may consider itself in the phase at a time.
>  * Ops will see enter() unblock in the order in which they tried to enter
>  * the phase.  entering (though not necessarily waiting for the future to
>  * resolve) a new phase prior to exiting the previous one will ensure that
>  * the op ordering is preserved.
>  */
> ```

I'll try this.

Reply:
> I think snaptrim_mutex has to be a member of PG, because the synchonization of SnapTrimEvents happens within the scope of PG. Am I missing anything?

No, you are right.

> ~I'll try this.~ It seems that the mutex has to be held after serveral pipeline phases, so it may not be feasible to share mutex with another pipeline phase.

I think we should either:
a) Aggregate the phases into a single one in order to use OrderedExclusivePhaseT logic (without the new mutex).
b) Simply lock and unlock when needed. (without introducing `WaitPG`).

Is there a reason I miss to both use a new `OrderedConcurrentPhaseT` + locking/unlocking?

Reply:
> a) Aggregate the phases into a single one in order to use OrderedExclusivePhaseT logic (without the new mutex).

This may not be feasible, because `SnapTrimEvents` have to be interleaved with `ClientRequests`; and if we use OrderedExclusivePhaseT, `ClientRequests` may have to wait for multiple `SnapTrimEvents` to complete before go ahead, which, I think, can significantly hurt `ClientRequests` throughput.

> b) Simply lock and unlock when needed. (without introducing `WaitPG`).

I added `WaitPG` so that `SnapTrimEvents` won't be in the previous exclusive phase when waiting for the lock, which may also hurt performance, Am I right? Thanks.

> 
> Is there a reason I miss that justifies using both a new `OrderedConcurrentPhaseT` + locking/unlocking?

Reply:
Why is it possible for multiple SnapTrimEvents to happen on the same PG concurrently?

Reply:
> Why is it possible for multiple SnapTrimEvents to happen on the same PG concurrently?

Every time a new snap remove happens, a new osdmap is created and a new snap trim event is created when activating new osdmap.

Reply:
No, OSDMap updates simply populates PG::snap_trimq.  As long as that set is non-empty, the snap trimming machine goes one snap at a time through that queue.  See PrimaryLogPG::SnapTrimmer.  I'd suggest using the same model here -- there's at most one fiber for each PG doing trimming (later, we'll want to limit how many are active concurrently) and OSDMap updates simply update a data structure and kick the fiber if it's not running.

Reply:
ok

Reply:
Though, the above seems like it might be an orthogonal change -- maybe it makes sense to merge the clones change for now and work on the snap trimming queue management change separately?

Reply:
> Though, the above seems like it might be an orthogonal change -- maybe it makes sense to merge the clones change for now and work on the snap trimming queue management change separately?

I agree that the management change can be separated from this PR.

Reply:
@athanatos, how does, if at all, classic guarantees the above change? IIUC the snap trimming machine manages the single snap handling at a time only.

Reply:
What do you mean by unfinished trimming state?  For each object that it trims, it holds the PG lock while constructing and submitting the repop.  The repop itself atomically removes the clone, updates the head metadata, and updates the SnapMapper stat via the normal replicated IO pathway.

I expect the difference with crimson would be the possibility of a concurrent IO updating PG state while we are doing IO as part of constructing the repop.  For normal IOs, the exclusive process phase prevents this.  Trimming will need to be in the same phase while constructing the repop for the same reasons as with classic.